zgw/common-ground-authentication/src/main/kotlin/nl/nlportal/commonground/authentication/CommonGroundAuthenticationConverter.kt

```diff
@@ -15,43 +15,87 @@
+                if (aanvrager is Map<*, *>) {
+                    if (aanvrager[BSN_KEY] != null) {
+                        return@map BurgerAuthentication(exchangedJwt, jwtGrantedAuthoritiesConverter.convert(exchangedJwt))
+                    } else if (aanvrager[KVK_NUMMER_KEY] != null) {
+                        return@map BedrijfAuthentication(exchangedJwt, jwtGrantedAuthoritiesConverter.convert(exchangedJwt))
+                    }
+                }
+
+                // This block is for temporary backwards compatibility
+                if (exchangedJwt.claims[BSN_KEY] != null) {
+                    return@map BurgerAuthentication(exchangedJwt, jwtGrantedAuthoritiesConverter.convert(exchangedJwt))
+                } else if (jwt.claims[KVK_NUMMER_KEY] != null) {
+                    return@map BedrijfAuthentication(exchangedJwt, jwtGrantedAuthoritiesConverter.convert(exchangedJwt))
+                }
+
+                val subject = exchangedJwt.subject
+                if (subject == null) {
+                    logger.error { "User with unknown subject has no bsn or kvk nummer assigned" }
+                } else {
+                    logger.error { "User with subject $subject has no bsn or kvk nummer assigned" }
+                }
+                 throw UserTypeUnsupportedException("User type not supported")
             }
         }
 
-        // This block is for temporary backwards compatibility
-        if (jwt.claims.get(BSN_KEY) != null) {
-            return Mono.just(BurgerAuthentication(jwt, jwtGrantedAuthoritiesConverter.convert(jwt)))
-        } else if (jwt.claims.get(KVK_NUMMER_KEY) != null) {
-            return Mono.just(BedrijfAuthentication(jwt, jwtGrantedAuthoritiesConverter.convert(jwt)))
-        }
 
-        val subject = jwt.subject
-        if (subject == null) {
-            logger.error { "User with unknown subject has no bsn or kvk nummer assigned" }
-        } else {
-            logger.error { "User with subject $subject has no bsn or kvk nummer assigned" }
-        }
-        throw UserTypeUnsupportedException("User type not supported")
     }
 
+    fun tokenExchange(jwt: Jwt): Mono<TokenResponse> {
+        return webClient.post()
+            .uri(URI.create("${jwt.issuer.toString().trimEnd('/')}/protocol/openid-connect/token"))
+            .body(
+                BodyInserters.fromFormData(
+                    LinkedMultiValueMap(
+                        mapOf(
+                            "client_id" to environment.getProperty("keycloak.resource"),
+                            "client_secret" to environment.getProperty("keycloak.credentials.secret"),
+                            "grant_type" to "urn:ietf:params:oauth:grant-type:token-exchange",
+                            "subject_token" to jwt.tokenValue,
+                            "requested_token_type" to "urn:ietf:params:oauth:token-type:access_token",
+                            "audience" to "gzac-portal-token-exchange",
```

Comment:
moet de audience niet gewoon uit de AZP claim gehaald worden net zoals bij de KeyCloakUserTokenExchangeFilter

Reply:
Helaas werkte dat niet vandaar deze aanpassing.
Wel nog van plan om te kijken of hier een generieke optie gemaakt kan worden als er meer tijd is nu moet dit morgen af.